src/gc/roots.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_GC_ROOTS_H
+#define PYSTON_GC_ROOTS_H
+
+#include "core/common.h"
+#include "core/threading.h"
+
+namespace pyston {
+
+#define GC_KEEP_ALIVE(t) asm volatile("" : : "x"(t))
+
+template <class T> class StackRoot {
+public:
+    explicit StackRoot(T* t) : t(t) {}
+    StackRoot(const StackRoot& other) : t(other.t) {}
+    template <class... Args> StackRoot(Args&&... args) : t(new T(args...)) {}
```

Comment:
I think this could use a ```t(new T(std::forward<Args>(args)...))``` but not sure if it makes a difference for the types we use it on.